### PR TITLE
Full systemd support

### PR DESCRIPTION
**Pull request by Kiskae**
**By submitting this pull request, I confirm the following:** 
*please fill any appropriate checkboxes, e.g: [X]*

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [ ] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
**What does this PR aim to accomplish?:**
Replaces the systemd service unit generated by sysv-generator with a custom one that manages the pihole-FTL service without using its daemon feature, since they seem to be incompatible. This resolves the restarting issue in #2869.

systemd output

**How does this PR accomplish the above?:**
* The service itself is a mirror of the actions defined in `/etc/init.d/pihole-FTL`, however instead of allowing the pihole binary to fork it uses the `no-daemon` flag to keep it in a systemctl-managed process.
* In order to share the code for initialization all the common code was split off into `/etc/pihole/fix_files.sh`, which is included in the sysv file using `source` and executed in systemd with `ExecStartPre=+` (to execute it as a privileged user).
* Running the daemon as an unprivileged user and granting of capabilities is done using native systemctl features instead of `setcap`/`su`.

**What documentation changes (if any) are needed to support this PR?:**
The change should be totally transparent to users, since it simply replaces the generated pihole-FTL.service with an actual implementation. The only exception is when a user has previously defined systemd overrides for the generated service. The automated install will detect and error out if these exist to avoid complication.

---
* You must follow the template instructions. Failure to do so will result in your pull request being closed.
* Please respect that Pi-hole is developed by volunteers, who can only reply in their spare time.

### CODE REVIEW

File: advanced/fix_files.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Touch files to ensure they exist (create if non-existing, preserve if existing)
+touch /var/log/pihole-FTL.log /var/log/pihole.log
+touch /run/pihole-FTL.pid /run/pihole-FTL.port
+touch /etc/pihole/dhcp.leases
+mkdir -p /var/run/pihole
+mkdir -p /var/log/pihole
```
Comment (dschaper) [inline advanced/fix_files.sh]: Can probably clean this up with a few changes.

```suggestion
mkdir -p /var/{run,log}/pihole
```

File: advanced/fix_files.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+chown "$USER": /var/run/pihole /var/log/pihole
+# Remove possible leftovers from previous pihole-FTL processes
+rm -f /dev/shm/FTL-* 2> /dev/null
```
Comment (dschaper) [inline advanced/fix_files.sh]: I don't think we need to redirect output to `/dev/null` as `rm -f` doesn't output anything if the files don't exist.

File: advanced/fix_files.sh [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+rm /var/run/pihole/FTL.sock 2> /dev/null
```
Comment (dschaper) [inline advanced/fix_files.sh]: Make this `rm -f` and we can lose the output redirect as well.

File: advanced/fix_files.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# Ensure that permissions are set so that pihole-FTL can edit all necessary files
+chown "$USER": /run/pihole-FTL.pid /run/pihole-FTL.port
+chown "$USER": /etc/pihole /etc/pihole/dhcp.leases 2> /dev/null
+chown "$USER": /var/log/pihole-FTL.log /var/log/pihole.log
```
Comment (dschaper) [inline advanced/fix_files.sh]: Could make these all `chown -f` and no need to redirect output.

File: advanced/fix_files.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Touch files to ensure they exist (create if non-existing, preserve if existing)
```
Comment (dschaper) [inline advanced/fix_files.sh]: Can we have a shebang on this? Probably should throw in the Pi-hole copyright header on it as well.
```suggestion
#!/usr/bin/env bash
# Pi-hole: A black hole for Internet advertisements
# (c) 2017 Pi-hole, LLC (https://pi-hole.net)
# Network-wide ad blocking via your own hardware.
#
# Completely uninstalls Pi-hole
#
# This file is copyright under the latest version of the EUPL.
# Please see LICENSE file for your rights under this license.
```

**Comment by dschaper** [comment]
The failed CI is because the PR is based off of master and it needs to be based off development with the latest CI updates. Please rebase your code off `development` and update the PR. Thanks!

File: advanced/fix_files.sh (hunk 5 above)
Comment (Kiskae) [reply advanced/fix_files.sh]: I've added the shebang and copyright header, although it has changed from the original where it sets the owner to `$USER` instead of `pihole`. For the modified `init.d/pihole-FTL` and the systemd definition it is correct but directly calling the file might produce unexpected behavior.

**Comment by Kiskae** [comment]
Did a rebase and force-push since the checks did not pick up that the base branch is now development, resulting in failures.

**Comment by dschaper** [comment]
This is a new error that I have not seen before. D-Bus issue it looks like?

https://travis-ci.org/pi-hole/pi-hole/builds/579268321#L965

**Comment by Kiskae** [comment]
Looks like its caused by the call to `systemctl daemon-reload` while `systemd` is not running. So for some reason in those containers there is a command `systemctl`, but `systemd` is not being used?

SU thread: https://superuser.com/questions/1173756/cant-start-systemd-failed-to-get-d-bus-connection

**Comment by dschaper** [comment]
There is no init system in the container itself, running systemd inside a docker container is not something that is done and not something CI/CD supports currently. @diginc I don't know if the situation has changed recently but I think we'd have to just mock the `systemctl` calls or intercept them?

**Comment by Kiskae** [comment]
In that case the correct solution is probably to move the `daemon-reload` call into the `main()` function, rather than `FTLinstall()`. If systemd is not running this does leave the configuration in a correct state.

**Comment by DL6ER** [comment]
@Kiskae I don't know if you are aware but this is not the first time we want to add a native systemd unit for `pihole-FTL`. I don't recall why we stopped back then but I like to see this coming up again.

There might be something interesting for you in https://github.com/pi-hole/pi-hole/pull/2112

**Comment by dschaper** [comment]
The previous attempt was starting to move to a systemd service file and abandon sysv init scripts. We cant do that so the thought was that if we need to continue to support sysv and the distros we support already have systemd/sysv helpers then there wasn't a need to make the systemd service file.

**Comment by DL6ER** [comment]
> The previous attempt was starting to move to a systemd service file and abandon sysv init scripts.

I just looked at the proposed code changes in #2112 and seem to disagree (although I see that the description is not accurate given the proposed changes). This PR tests for `systemd` not only being present but also being the used `init` system. Only in this case, we installed the systemd unit (but not the sysv script). If the `init` system was not `systemd` nothing would have been changed to how we do it now.

File: advanced/Templates/pihole-FTL.service.ini [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+After=resolvconf.service
+Wants=resolvconf.service
+# fail since dnsmasq is embedded
+Conflicts=dnsmasq.service
+
+[Service]
+# Run as unprivileged user, use AmbientCapabilities to assign capabilities
+User=pihole
+Group=pihole
+
```
Comment (DL6ER) [inline advanced/Templates/pihole-FTL.service.ini]: Shouldn't we add
```suggestion
Type=forking
PIDFile=/var/run/pihole-FTL.pid
```
?

File: advanced/Templates/pihole-FTL.service.ini [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+# Assign capabilities to the pihole user slice
+AmbientCapabilities=CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_NET_ADMIN
+ExecStart=/usr/bin/pihole-FTL no-daemon
+
+## stop()
+# Clean up resolv.conf before shutting down service
+ExecStop=+/sbin/resolvconf -d lo.piholeFTL
+# Send SIGKILL 5 seconds after SIGTERM
+TimeoutStopSec=5
+
```
Comment (DL6ER) [inline advanced/Templates/pihole-FTL.service.ini]:
```suggestion
ExecReload=/bin/kill -HUP $MAINPID
```

File: advanced/Templates/pihole-FTL.service.ini [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ExecStartPre=+/bin/sh -c "echo \"nameserver 127.0.0.1\" | /sbin/resolvconf -a lo.piholeFTL"
+# Assign capabilities to the pihole user slice
+AmbientCapabilities=CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_NET_ADMIN
+ExecStart=/usr/bin/pihole-FTL no-daemon
+
+## stop()
+# Clean up resolv.conf before shutting down service
+ExecStop=+/sbin/resolvconf -d lo.piholeFTL
+# Send SIGKILL 5 seconds after SIGTERM
+TimeoutStopSec=5
```
Comment (DL6ER) [inline advanced/Templates/pihole-FTL.service.ini]: We might want to give FTL a bit more time. After receiving `SIGTERM`, it may still have to write a lot of history to the database. In a previous PR, I proposed a limit of 10 seconds here.
```suggestion
TimeoutStopSec=10
```

File: advanced/Templates/pihole-FTL.service.ini [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+User=pihole
+Group=pihole
+
+# start()
+ExecStartPre=+/bin/sh /etc/pihole/fix_files.sh
+ExecStartPre=+/bin/sh -c "echo \"nameserver 127.0.0.1\" | /sbin/resolvconf -a lo.piholeFTL"
+# Assign capabilities to the pihole user slice
+AmbientCapabilities=CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_NET_ADMIN
+ExecStart=/usr/bin/pihole-FTL no-daemon
+
```
Comment (DL6ER) [inline advanced/Templates/pihole-FTL.service.ini]: Previously, I proposed
```suggestion
; Make /usr, /boot, /etc and possibly some more folders read-only...
ProtectSystem=full
; ... except /etc/pihole
; This merely retains r/w access rights, it does not add any new.
; Must still be writable on the host!
ReadWriteDirectories=/etc/pihole
```
Do you think this is a good idea?

**Comment by dschaper** [comment]
Thanks for the link, I've reviewed the thread and I think the general tone of things was, "We can do both, but it's going to be hard and problematic." So I'll link a comment here that may be good to review. https://github.com/pi-hole/pi-hole/pull/2112/files#r189458307

We keep sysvinit for the obscure installs like Synology or chroot'ed servers. A year ago this was a need. I don't think we need to anymore. @diginc has an amazing docker image now that can run in these cases.

File: advanced/Templates/pihole-FTL.service.ini [hunk 10]
```diff
@@ -0,0 +1,37 @@
+[Unit]
+Description=pihole-FTL Domain Name Server
+Documentation=man:pihole-FTL(8)
+After=remote-fs.target
+
+# Alternative based on https://wiki.debian.org/Hardening/Daemons#bind
+#   allows other services which rely on DNS to use pihole-FTL, but means
+#   breaking if config relies on remote-fs
+# After=network.target
```
[the remark this answers is not in the document]
Comment (Kiskae) [reply advanced/Templates/pihole-FTL.service.ini]: Its disabled because I'm not entirely sure what the interactions between `After=remote-fs.target` and `After=network.target` would be, so I just left the dependencies as a mirror of the init script.

File: advanced/Templates/pihole-FTL.service.ini (hunk 6 above)
Comment (Kiskae) [reply advanced/Templates/pihole-FTL.service.ini]: No we're running pihole-FTL in no-daemon mode, since FTL's double-fork daemon implementation actually seems to cause issues. So the implied `Type=simple` is used where systemd maintains direct control over the process.

File: advanced/Templates/pihole-FTL.service.ini (hunk 8 above)
Comment (Kiskae) [reply advanced/Templates/pihole-FTL.service.ini]: It would probably be better as a separate PR since you'd want to change the init script at the same time?

File: advanced/Templates/pihole-FTL.service.ini (hunk 9 above)
Comment (Kiskae) [reply advanced/Templates/pihole-FTL.service.ini]: It looks like this might break existing setups if people changed their configs: 
[Configurable paths](https://github.com/pi-hole/FTL/blob/9e8273b748a6a0870c64fcc0ec4880a28bbcd587/config.c#L282-L310)

It might be interesting as an optional security feature which could be installed as a drop-in config like `/etc/systemd/system/pihole-FTL.service.d/protect-fs.conf`

File: advanced/Templates/pihole-FTL.service.ini (hunk 7 above)
Comment (Kiskae) [reply advanced/Templates/pihole-FTL.service.ini]: Like with the timeout change, the init script currently does a full restart on `service pihole-FTL reload`, so it should probably involve a change to https://github.com/pi-hole/pi-hole/blob/master/advanced/Templates/pihole-FTL.service#L103 at the same time?

File: advanced/Templates/pihole-FTL.service.ini (hunk 10 above)
[the remark this answers is not in the document]
Comment (Kiskae) [reply advanced/Templates/pihole-FTL.service.ini]: After thinking about it some more, it should be fine since multiple `After`'s are safe. The problem would be between `After=remote-fs.target` and `Before=nss-lookup.target` since presumably remote mounts would require DNS in some cases.

File: advanced/Templates/pihole-FTL.service.ini (hunk 7 above)
Comment (DL6ER) [reply advanced/Templates/pihole-FTL.service.ini]: Yes.

File: advanced/Templates/pihole-FTL.service.ini (hunk 8 above)
Comment (DL6ER) [reply advanced/Templates/pihole-FTL.service.ini]: Yes, we do this separately.

File: advanced/Templates/pihole-FTL.service.ini (hunk 9 above)
Comment (DL6ER) [reply advanced/Templates/pihole-FTL.service.ini]: Ah yes, I basically forgot about the possibility to move all the files. In this case, we probably shouldn't do it.

**Comment by Kiskae** [comment]
I believe it might be outside of the scope of this PR since my goal was to fix compatibility with debian buster by fixing the execution through systemd (#2869) by creating a systemd service unit with the same behavior as the existing init script.
Completely swapping over to a different system manager or making changes to the existing behavior would probably be better as a separate PR in my opinion.

**Comment by dschaper** [comment]
I agree with that.

**Comment by Kiskae** [comment]
I've made two last adjustments to the PR:

1. I've moved `fix_files.sh` to `/opt/pihole`, since `/etc` should be used for configuration, not executable code.
2. Added `network.target` as an ordering dependency as per the comments by @DL6ER, I believe adding further dependencies for `nss-lookup.target` would be risky since it could potentially cause the following cycle:
    `pihole-FTL.service -> remote-fs.target -> (mounting service) -> nss-lookup.target -> pihole-FTL.service` if a networking file system solution uses DNS.

**Comment by MichaIng** [comment]
Probably instead of `network.target`, it should be `network-online.target` to solve the following issue: https://github.com/pi-hole/pi-hole/issues/2924
But could be tested if `network.target` is sufficient already. Depends on how exactly `dnsmasq` derives info about requests are from local network or not.

And it makes sense then to add `$network` to the sysvinit service as well, which leads to `network-online.target` in the systemd translator. But see the linked issue.
___________
Another topic when already touching the services:
- What was the reason to add `$syslog` to the startup dependencies?
- Since Pi-hole has it's own log files, this should be not required and on systemd, the sysvinit => systemd translator ignores it anyway. So IMO it can be removed from the sysvinit service, although it does not hurt as well.
- On DietPi, we removed it/replaced it with `$network`, since there have been some issues with Pi-hole startup if `rsyslog` was not installed or not active in the past. But I could not replicate on Debian up from Stretch at least. And since `pihole -up` overrides the service script, this step does not make sense anyway 😉.

**Comment by Kiskae** [comment]
Regarding `network-online.target` it would probably fall under the same reasoning as the change to reload and the stop timeout in that it involves a change to the sys-v script as well, so its probably best left as a separate PR.

However I did end up discovering that `network.target` only really has an effect during system shutdown ([NetworkTarget](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/)) where connections should be gracefully shut down before the network adapters are shut down. So I ended up creating the following list of potentially relevant dependencies and how they affect the pihole-FTL unit:

- `remote-fs.target` - Ensures that any configuration files present on remote file systems are fully available before the service starts
  - Could be replaced with `RequiresMountsFor=/etc/pihole,/opt/pihole,/run,/var/log` so that systemd automatically links the required mounting targets, but this would break if the configuration was customized (see https://github.com/pi-hole/pi-hole/pull/2900#discussion_r319773330)
- `network.target` - Ensures connections are gracefully shut down, instead of abruptly turning off the network adapter.
- `network-online.target` - To maintain parity with sys-v would require addition of `$network` (see https://github.com/pi-hole/pi-hole/pull/2900#issuecomment-531490783). While the exact requiremetns for this target are vague it is implied that at least one adapter is fully configured (potentially through DHCP, which would explain why it sometimes takes a while to fully start)
- `nss-lookup.target` - This dependency would be useful if you want to force any DNS lookups the system does through pihole, but since its dynamically enabled through `resolvconf` this is not a hard dependency and would probably cause more potential issues than it solves.
- `resolvconf.service` - This dependency is required since it toggles the mutability of the `resolvconf` config file, if it isn't started before pihole it would fail to add the local nameserver.
- `dnsmasq.service` - Since `dnsmasq` is embedded this should be a simple `Conflicts` dependency unless there is a usecase for running both dnsmasq and pihole-FTL on a machine.

In conclusion the changes mentioned by @MichaIng are probably outside of the scope of this PR since they involve fixing the sys-v script as well. In addition both `network.target` and `network-online.target` would be required since they have different implications for the behavior of the service during shutdown.

**Comment by dschaper** [comment]
`nss-lookup.target` and `resolvconf.service` are not needed in my opinion. We've talked about dropping the hardcode of localhost for the system resolver of the Pi-hole node. That decision should be left up to the user and really only is helpful if you are running with a head on the Pi-hole and doing browsing that would need sinkhole capabilities.

That said, if it would be better for this PR to just duplicate the existing functionality and then make changes to not use localhost in a new PR then I'm good with that.

**Comment by MichaIng** [comment]
@Kiskae 
Just as little hint:
```
root@DietPi:~# systemctl cat network-online.target
# /lib/systemd/system/network-online.target
#  SPDX-License-Identifier: LGPL-2.1+
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=Network is Online
Documentation=man:systemd.special(7)
Documentation=https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget
After=network.target
```
- `network-online.target` is only reached after `network.target`, thus reversed on shutdown. So when adding `After=network-online.target`, adding `network.target` is never required.

However yes, probably these things are better handled in a separate PR/topic. Only concern is, that for debug and end user support reasons, it is easier to have both services, sysvinit and systemd, behave as similar as possible.

File: advanced/fix_files.sh [hunk 11]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+# Pi-hole: A black hole for Internet advertisements
+# (c) 2017 Pi-hole, LLC (https://pi-hole.net)
+# Network-wide ad blocking via your own hardware.
+#
+# Completely uninstalls Pi-hole
```
Comment (Mcat12) [inline advanced/fix_files.sh]: This description should be updated.

File: advanced/fix_files.sh [hunk 12]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+# Pi-hole: A black hole for Internet advertisements
+# (c) 2017 Pi-hole, LLC (https://pi-hole.net)
```
Comment (Mcat12) [inline advanced/fix_files.sh]: Copyright year can be updated to 2019

**Comment by PromoFaux** [comment]
@Kiskae, sorry to be a pain, please review and fix the merge conflicts :)

**Comment by Kiskae** [comment]
@PromoFaux Its been rebased on upstream `development`, seems no other fixes were needed.

**Comment by dschaper** [comment]
Version 5 has some changes that affect this. Localhost is no longer injected to resolvconf, nor is localhost required for FTL operation. Closing this PR as it's not compatible, but not closing the idea. A new PR based on the release/v5.0 branch is welcome.

**Comment by MichaIng** [comment]
@dschaper 
Why not leaving this open to be adjusted for v5 compatibility? No need to redo the 95% of this PR which are still valid. I agree that the service itself should not touch `resolvconf`/`resolv.conf` at all. This should still be the job of at best the installer and afterwards the admin, as there might be custom setups where it is not wanted to have the server itself resolving through Pi-hole and/or an own DNS node is running on localhost 53 port or such.

@Kiskae 
Only if you agree to do required changes of course.
One note about `is_command systemctl`:
- The convention to check whether systemd is the init system is to check for `[[ -d '/run/systemd/system' ]]`.
- The above should be used to decide whether to enable+start / disable+stop the systemd unit.
- However the systemd unit should be installed (but not enabled+started) in any case, so it is present in case the system is upgraded, migrated from sysvinit to systemd or such.
- This would follow usual package conventions.

**Comment by dschaper** [comment]
I think the major changes in v5 warrant a new PR with a new review of what is needed. It's stable enough for integration testing with DietPi to see what effect our changes have.

File: automated install/basic-install.sh [hunk 13]
```diff
@@ -2207,8 +2207,23 @@ FTLinstall() {
...
-    # Always replace pihole-FTL.service
-    install -T -m 0755 "${PI_HOLE_LOCAL_REPO}/advanced/Templates/pihole-FTL.service" "/etc/init.d/pihole-FTL"
+    # Ensure /opt/pihole exists, then install fix_files.sh
+    install -T -D -m 0755 "${PI_HOLE_LOCAL_REPO}/advanced/fix_files.sh" "${PI_HOLE_INSTALL_DIR}/fix_files.sh"
+    # Always replace pihole-FTL
+    install -T -m 0755 "${PI_HOLE_LOCAL_REPO}/advanced/Templates/pihole-FTL.sh" "/etc/init.d/pihole-FTL"
+    
+    if is_command systemctl ; then
+        local systemd_unit_file="/etc/systemd/system/pihole-FTL.service"
+        if [[ ! -f "${systemd_unit_file}" && -d "${systemd_unit_file}.d" ]]; then
```
Comment (MichaIng) [inline automated install/basic-install.sh]: I would not error out here since there are plenty of other reasons to use override files, e.g. service hardening or nice/priority levels, Before=/After= ordering and all such. But a warning would be fine instead, so users are reminded to check back if anything fails.

**Comment by MichaIng** [comment]
@dschaper 
Okay, probably I find time to do one, or @Kiskae is faster since I actually have another thing that I wanted to contribute before I forget it. Failsafe check for which APT packages are available:
- Issue: #2888 
- PR that included the fix: https://github.com/pi-hole/pi-hole/pull/2889/files#diff-689793e68cde6270d074695b71c969afR208-R255

**Comment by MichaIng** [comment]
@dschaper 
DietPi support btw is done with: https://github.com/MichaIng/DietPi/commit/b90b49fe701546da69f87b10de6f70e6009b5cb0#diff-d92a6ee04e02fd2a2dc23d5bec3e6a98R8767
There was happily not much todo and I added an error handler around it. The Pi-hole installer adds `pihole` user to `www-data` group as well, but from what I see this has not really an effect since blocking page and admin  dirs are owned by root with default 755/644 permissions.

While I would love to see a systemd unit, it is nothing urgend and not related to DietPi-compatibility at all :slightly_smiling_face:.

File: automated install/basic-install.sh (hunk 13 above)
Comment (Kiskae) [reply automated install/basic-install.sh]: That check exists for cases where override files were created to work with the generated system-v units, so it would error out to prevent unintended effects from combining those with a new systemd definition. It might still be a good idea to notify the user if the systemd file is updated and existing overrides exist.

**Comment by Kiskae** [comment]
I don't actually have a pi-hole setup to work with at the moment, so feel free to develop an updated patch based on the work done in this PR.

File: automated install/basic-install.sh (hunk 13 above)
Comment (MichaIng) [reply automated install/basic-install.sh]: Yes I understood the reason for this check, but as said, it breaks the install when there are non-evil override files for several other good reasons. IMO it would be better to assume it's non-harmful and warn/remind users to check it in case of errors, instead of forcing all to move all overrides out of place, run the install, move overrides into place again, restart service... and probably then the installer fails because for some special permissions/capabilities on `pihole-FTL.service` the override is required.

**Comment by pralor** [comment]
This pull request has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/have-to-pihole-restartdns-after-reboot/28772/21

**Comment by dschaper** [comment]
>  The Pi-hole installer adds `pihole` user to `www-data` group as well, but from what I see this has not really an effect since blocking page and admin dirs are owned by root with default 755/644 permissions.

I think you will find that the group membership is needed.

**Comment by MichaIng** [comment]
@dschaper 
```
2020-03-10 18:52:36 root@VM-Buster:~# find / -group www-data | grep -v '^/proc/'
/run/lighttpd
/run/lighttpd/php.socket-0
/var/log/lighttpd
/var/log/lighttpd/access.log
/var/log/lighttpd/error.log
/var/www/html
/var/cache/lighttpd
/var/cache/lighttpd/uploads
/var/cache/lighttpd/compress
/etc/pihole/migration_backup/regex.list
2020-03-10 18:55:42 root@VM-Buster:~# l /run/lighttpd /var/log/lighttpd /var/cache/lighttpd /etc/pihole/migration_backup/regex.list
-rw-rw-r-- 1 pihole   www-data    0 Mar 10 18:47 /etc/pihole/migration_backup/regex.list

/run/lighttpd:
total 0
srwxr-xr-x 1 www-data www-data 0 Mar 10 18:49 php.socket-0

/var/cache/lighttpd:
total 8.0K
drwxr-x--- 2 www-data www-data 4.0K Apr 13  2019 compress
drwxr-x--- 2 www-data www-data 4.0K Apr 13  2019 uploads

/var/log/lighttpd:
total 4.0K
-rw-r--r-- 1 www-data www-data   0 Mar 10 18:47 access.log
-rw-r--r-- 1 www-data www-data 357 Mar 10 18:49 error.log
```
Does the `pihole-FTL` read Lighttpd uploads or compress dir directly in some case? Otherwise as can be seen, the group membership adds no other ability.

**Comment by dschaper** [comment]
Look at `/etc/pihole/` or `/etc/dnsmasq.d/*`

**Comment by dschaper** [comment]
Group memberships were not set just willy-nilly. There's a specific reason for it and for the security of running `pihole` commands via the web interface. You can remove those features from your distribution if you'd like but I think you'll find them to be required once your users find things breaking.

**Comment by MichaIng** [comment]
@dschaper 
```
2020-03-10 18:56:09 root@VM-Buster:~# l /etc/pihole/ /etc/dnsmasq.d/*
-rw-r--r-- 1 root   root   1.4K Mar 10 18:49 /etc/dnsmasq.d/01-pihole.conf

/etc/pihole/:
total 18M
-rw-r--r-- 1 root   root    313 Mar 10 18:46 adlists.list.old
-rw-r--r-- 1 pihole pihole    0 Mar 10 18:49 dhcp.leases
-rw-r--r-- 1 root   root    596 Mar 10 18:49 dns-servers.conf
-rw-r--r-- 1 root   root     13 Mar 10 18:48 ftlbranch
-rw-r--r-- 1 root   root     18 Mar 10 18:49 GitHubVersions
-rw-rw-r-- 1 pihole pihole 9.1M Mar 10 18:49 gravity.db
-rw-r--r-- 1 root   root   2.8M Mar 10 18:48 gravity.list
-rw-r--r-- 1 root   root   1.1K Mar 10 18:49 install.log
-rw-r--r-- 1 root   root   995K Mar 10 18:49 list.0.raw.githubusercontent.com.domains
-rw-r--r-- 1 root   root   581K Mar 10 18:49 list.1.mirror1.malwaredomains.com.domains
-rw-r--r-- 1 root   root   403K Mar 10 18:49 list.2.sysctl.org.domains
-rw-r--r-- 1 root   root    521 Mar 10 18:49 list.3.s3.amazonaws.com.domains
-rw-r--r-- 1 root   root    43K Mar 10 18:49 list.4.s3.amazonaws.com.domains
-rw-r--r-- 1 root   root   1.3M Mar 10 18:49 list.5.hosts-file.net.domains
-rw-r--r-- 1 root   root     38 Mar 10 19:30 localbranches
-rw-r--r-- 1 root   root     48 Mar 10 18:49 local.list
-rw-r--r-- 1 root   root     52 Mar 10 19:30 localversions
-rw-r--r-- 1 root   root    234 Mar 10 18:49 logrotate
-rw-r--r-- 1 pihole pihole 2.3M Mar 10 18:49 macvendor.db
drwxr-xr-x 2 root   root   4.0K Mar 10 18:49 migration_backup
-rw-rw-r-- 1 pihole root     15 Mar 10 18:49 pihole-FTL.conf
-rw-r--r-- 1 pihole pihole  44K Mar 10 19:30 pihole-FTL.db
-rw-r--r-- 1 root   root    300 Mar 10 18:49 setupVars.conf
-rw-r--r-- 1 root   root    300 Mar 10 18:49 setupVars.conf.update.bak
```
I see the reasons here why `www-data` must be a member of the `pihole` group (already added to DietPi), but not the other way round: There is nothing owned by the `pihole` group here, besides the above listed `regex.list` where `pihole` user is already the owner, hence does not benefit from `www-data` group.

> Group memberships were not set just willy-nilly.

This is the reason why I am digging into this instead of just adding it to our code 😉. The less memberships, the more secure.
________
Let me check where and why this group membership was added...

**Comment by MichaIng** [comment]
Okay tracked down and this group membership was already part of the very first installer script commit when the `pihole` did not even exist: https://github.com/pi-hole/pi-hole/blob/e8b33ec8eef4631bab75a3ce07b4c231bc395f2c/advanced/automated%20instal/basic-install.sh
The `pihole` user was implemented first as fallback if `pi` does not exist to enable wider support non-RPi systems. The `pi` user was and is a semi-admin user as it is by default part of all relevant system groups, including `www-data`. Sadly there is no real explanation why the membership is redone.

@jacobsalmela
Do you remember why added `usermod -a -G www-data pi` to the initial installer script that time?
